### PR TITLE
#549 fix URL equals

### PR DESCRIPTION
**Pull request by Emily-Jiang**
Signed-off-by: Emily Jiang <emijiang6@googlemail.com>

### CODE REVIEW

**Comment by Emily-Jiang** [comment]
@radcortez please review!

**Comment by Emily-Jiang** [comment]
@jbee please review

File: tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java [hunk 1]
```diff
@@ -1654,120 +1658,161 @@ public void testInstantSetInjection() {
...
+                return expectedValue[i].toURI().equals(value[i].toURI());
+        } 
+        catch (URISyntaxException e) {
+                e.printStackTrace();
+        }
+        return false;
+        }));
+    }
+
+    private void assertURLListEquals(List<URL> value) throws MalformedURLException {
```
Comment (jbee) [inline tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java:1677]: I'd think that either of the `assertURLArrayEquals` and `assertURLListEquals` could be implemented in terms of the other. I'd do `assertURLArrayEquals` => `assertURLListEquals(asList(value))`

File: tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java [hunk 2]
```diff
@@ -1654,120 +1658,161 @@ public void testInstantSetInjection() {
...
+        }));
+    }
+
+
     @Test
-    public void testUrlArrayLookupProgrammatically() throws MalformedURLException {
+    public void testUrlArrayLookupProgrammatically() throws MalformedURLException, URISyntaxException {
         URL[] value = config.getValue("tck.config.test.javaconfig.converter.urlvalues", URL[].class);
         Assert.assertNotNull(value);
         Assert.assertEquals(value.length, 3);
```
Comment (jbee) [inline tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java:1698]: If there is a special method to assert equality I'd think the length and null check can or should be part of it so they do not need duplication.

File: tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java [hunk 3]
```diff
@@ -1654,120 +1658,161 @@ public void testInstantSetInjection() {
...
-        Assert.assertEquals(value, new URL[]{
-            new URL("http://microprofile.io"),
-            new URL("http://openliberty.io"),
-            new URL("http://microprofile.io")});
+        assertURLArrayEquals(value);
         URL[] single = config.getConverter(URL[].class).get().convert("http://microprofile.io");
         Assert.assertNotNull(single);
         Assert.assertEquals(single.length, 1);
-        Assert.assertEquals(single, new URL[]{ new URL("http://microprofile.io") });
+        Assert.assertEquals(single[0].toURI(), new URL[]{ new URL("http://microprofile.io") }[0].toURI());
```
Comment (jbee) [inline tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java:1715]: Similarly to the extracted method to assert an array with 3 elements it seems worth it to extract one for the array with single element (including the null and length check) as it is a reoccurring assertion pattern.

File: tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java [hunk 4]
```diff
@@ -1654,120 +1658,161 @@ public void testInstantSetInjection() {
...
-            new URL("http://openliberty.io"),
-            new URL("http://microprofile.io")));
+        assertURLListEquals(values);
         Optional<List<URL>> optionalSingle = config.getOptionalValues("tck.config.test.javaconfig.converter.urlvalue", URL.class);
         Assert.assertTrue(optionalSingle.isPresent());
         List<URL> single = optionalSingle.get();
         Assert.assertNotNull(single);
         Assert.assertEquals(single.size(), 1);
-        Assert.assertEquals(single, Arrays.asList(new URL("http://microprofile.io")));
+        Assert.assertEquals(single.get(0).toURI(), Arrays.asList(new URL("http://microprofile.io")).get(0).toURI());
```
Comment (jbee) [inline tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java:1759]: `Arrays.asList(new URL("http://microprofile.io")).get(0)` should just be `new URL("http://microprofile.io")`

File: tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java [hunk 5]
```diff
@@ -1654,120 +1658,161 @@ public void testInstantSetInjection() {
...
-            new URL("http://microprofile.io"),
-            new URL("http://openliberty.io"),
-            new URL("http://microprofile.io")});
+    public void testUrlArrayInjection() throws MalformedURLException, URISyntaxException {
+        URL[] values = converterBean.getMyUrlArray();
+        Assert.assertEquals(values.length, 3);
+        assertURLArrayEquals(values);
         Assert.assertEquals(converterBean.getMySingleUrlArray().length, 1);
-        Assert.assertEquals(converterBean.getMySingleUrlArray(), new URL[]{ new URL("http://microprofile.io") });
+        Assert.assertEquals(converterBean.getMySingleUrlArray()[0].toURI(), new URL[]{ new URL("http://microprofile.io") }[0].toURI());
```
Comment (jbee) [inline tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java:1769]: Equally `new URL[]{ new URL("http://microprofile.io") }[0]` => `new URL("http://microprofile.io")`

File: tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java [hunk 6]
```diff
@@ -1654,120 +1658,161 @@ public void testInstantSetInjection() {
...
-    public void testURLListInjection() throws MalformedURLException {
-        Assert.assertEquals(converterBean.getMyUrlList().size(), 3);
-        Assert.assertEquals(converterBean.getMyUrlList(), Arrays.asList(
-            new URL("http://microprofile.io"),
+    public void testURLListInjection() throws MalformedURLException, URISyntaxException {
+        List<URL> values = converterBean.getMyUrlList();
+        Assert.assertEquals(values.size(), 3);
+        assertURLListEquals(values);
+        Assert.assertEquals(converterBean.getMySingleUrlList().size(), 1);
+        Assert.assertEquals(converterBean.getMySingleUrlList().get(0).toURI(), Arrays.asList(new URL("http://microprofile.io")).get(0).toURI());
```
Comment (jbee) [inline tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java:1778]: unnecessary list wrapping again, maybe it is intentional? I find it more confusing.

File: tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java [hunk 7]
```diff
@@ -1654,120 +1658,161 @@ public void testInstantSetInjection() {
...
+    }
+
+    private boolean assertURLSetEquals(Set<URL> valueSet, Set<URL> expectedURLSet) throws MalformedURLException, URISyntaxException {
+        if (valueSet.size()!= expectedURLSet.size()) {
+            return false;
+        }
+        Iterator<URL> it = valueSet.iterator();
+        boolean isEquals = true;
+        while (it.hasNext()) {
```
Comment (jbee) [inline tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java:1787]: Using a `while` and some `for` loops seems overly complicated and not very readable. I suggest to map to sets of URIs and use `uriSet1.equals(uriSet2)`

File: tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java [hunk 8]
```diff
@@ -1654,120 +1658,161 @@ public void testInstantSetInjection() {
...
-        Assert.assertEquals(converterBean.getMySingleUrlList(), Arrays.asList(new URL("http://microprofile.io")));
-    }
-
-    @Test
-    public void testURLSetInjection() throws MalformedURLException {
-        Assert.assertEquals(converterBean.getMyUrlSet().size(), 2);
-        Assert.assertEquals(converterBean.getMyUrlSet(), new LinkedHashSet<>(Arrays.asList(
-                new URL("http://openliberty.io"),
-                new URL("http://microprofile.io"))));
+        Assert.assertTrue(assertURLSetEquals(values, expectedURLSet));
```
Comment (jbee) [inline tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java:1811]: As sets are used here I wonder if the lists and arrays used before are intentionally expecting an order or if those really should compare sets as well. Can say but wanted to point it out. 
Using a `LinkedHashSet` also bags the question if order does matter in which case these maybe should be compared as lists? Or use the default `HashSet` to leave no question that these are sets and order is not important.

File: tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java (hunk 2 above)
Comment (Emily-Jiang) [reply tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java:1698]: I think this is a short circuit. The whole test will be failed if the value is null or contains a different size of the items.

File: tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java (hunk 3 above)
Comment (Emily-Jiang) [reply tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java]: ok. I will refactor.

File: tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java (hunk 8 above)
Comment (Emily-Jiang) [reply tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java:1834]: not really. We should not force ordering for sets.

File: tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java (hunk 6 above)
Comment (Emily-Jiang) [reply tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java]: refactored.

File: tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java (hunk 7 above)
Comment (Emily-Jiang) [reply tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java:1810]: What you suggested will be equivalent URI comparison and defeat the purpose of the URL comparison, which means the set contains a collection of URLs.

Comment (Emily-Jiang) [reply tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java:1810]: Most important thing is to test the right thing.

**Comment by Emily-Jiang** [comment]
@jbee I have done most things you suggested and added explaintation for the comments I did not do. Please approve if you have no further comments.

File: tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java (hunk 8 above)
Comment (jbee) [reply tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java:1834]: > We should not force ordering for sets.

Then why use a `LinkedHashSet`?

File: tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java (hunk 7 above)
Comment (jbee) [reply tck/src/main/java/org/eclipse/microprofile/config/tck/ArrayConverterTest.java:1810]: The code already does `thisURL.toURI().equals(url.toURI())` (URI comparison) and is basically just a overly complicated way to compare two sets of URL in terms of their URI equality. Or I misread this code entirely. Anyhow. This was just a suggestion to make the code more readable.

**Comment by Emily-Jiang** [comment]
Thanks @jbee !

**Pull request merged**